Ignore collisions for named ports that are not actually used in an egress policy

jrajahalme opened the pull request:
Do not warn of port name conflicts when ports are collected, but keep
the conflicting ports. This allows (any) warnings to be postponed until
the named ports are used in a policy.

Behavior in presence of conflicting port name/number mapping, when
the port name is used in a policy is slightly changed:

 - Now these collisions will not affect ingress policies at all
 - For egress the policy rule using the conflicting port is now ignored.
   Cilium needs to be running with debug on to see this happening.
   Previously rule was instantiated using one of the many port numbers,
   essentially selected randomly.
 - Missing port name mappings, when used in a policy applicable to
   the local node, are logged as warnings, but rate limited.

Pass Endpoint (or a subset interface) to potential users of Endpoint's
NamedPortsMap so that instead of preparing the map in advance it can
be created of demand in case it is actually needed.

NamedPortsMap is an interface that is implemented by both NamedPortMap
or NamedPortMultiMap, first of which is built from the POD spec, and
the latter is a collection of named ports across all the PODs in the
cluster. When a named port is used in an ingress policy, it can be
translated by using the PODs own NamedPortMap. When a named port is
used in an egress policy the named port can be one of any remote PODs,
so the collection of all the named ports in the cluster is used as the
individual destination POD is not known at policy computation
time. This could be optimized to the scope of PODs sharing the same
set of labels as the destination security IDs are known when policy is
configured.

Currently same port name resolving to different port numbers is
flagged as a runtime error when that specific port name is used in a
policy. Traffic to this/these ports is not allowed. To get around this the
port names should be made unique so that each name maps to the
same port number and protocol, or preferably the port names in the
egress policies should be replaced with the desired port numbers.

The maintenance of the collection of named ports across the cluster
described above is started only after first needed for computing the
bpf map entries from an egress policy that uses a port name instead
the number. This way the related overhead is avoided if named ports
are never used in egress policies.

Fixes: #12514

jrajahalme committed:
ipcache: Keep port name conflicts

Do not warn of port name conflicts when ports are collected, but keep
the conficting ports. This allows (any) warnings to be postponed until
the named ports are used in a policy.

policy.NamedPortsMap is renamed to policy.NamedPortMap, and a new
policy.NamedPortMultiMap is defined, as well as an interface
policy.NamedPortsMap that abstracts the difference for callers of
GetNamedPort().

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/proxy/mock_test.go:
```diff
@@ -24,7 +24,7 @@ import (
 )
 
-type proxyUpdaterMock struct {
+type ProxyUpdaterMock struct {
```
If we're duplicating this in `pkg/envoy` and `pkg/proxy`, maybe we should place a single version under `pkg/testutils` instead?

jrajahalme replied:
Tried placing this in `pkg/testutils`, but unfortunately that introduced an import loop. Also, this will be removed from `pkg/proxy` by https://github.com/cilium/cilium/pull/12548.

joestringer replied:
SGTM. I wonder if a new `pkg/testutils/proxy` package would make sense for this but if there's only going to be one instance soon anyway then the point is moot.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -44,20 +44,38 @@ import (
 )
 
+// GetNamedPortsMap returns the map of named ports relevant for the given direction
+func (e *Endpoint) GetNamedPortsMap(ingress bool) policy.NamedPortsMap {
```
I assume that the locking aspect is OK here because we return a map by value so there's no way the caller could interfere with the actual `e.k8sPorts` copy of the map?

jrajahalme replied:
Maps are always returned by reference, so here we trust that the caller is not messing with the returned map.

christarazi replied:
Should we return a deep copy of the map in that case then?

joestringer replied:
So we don't actually have any locking guarantee here then? Shouldn't we at least readlock while reading this?

aanm replied:
@christarazi as long we document that the caller should not mess up with the map, it should be fine. We shouldn't do useless deepcopies.

+1 for the missing locking guarantee that Joe is pointing out

joestringer reviewed (approved):
I didn't see at a glance how we continue to warn about collisions later on when implementing the policy, but I'm guessing that's because it's existing code somewhere that's not touched by this PR.

But otherwise looks good, couple of minor points below but could likely be deferred for subsequent cleanup.

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -211,7 +211,7 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
 				var revertFunc revert.RevertFunc
 
-				proxyID, err := e.ProxyID(e.desiredPolicy.NamedPortsMap, l4)
+				proxyID, err := e.ProxyID(l4)
```
Typo is commit msg:

"of demand" -> "on demand"

jrajahalme replied:
Fixed, thanks!

christarazi commented on pkg/envoy/mock_test.go:
```diff
@@ -0,0 +1,65 @@
+// Copyright 2017-2019 Authors of Cilium
```
This file doesn't have any tests in it. Is there a reason it is named "mock_test"? Could it just be "mock.go"?

christarazi replied:
We could also introduce a `mock` subpackage inside the `envoy` package, that way we don't pollute the `envoy` package namespace and reduce the burden of importing it.

christarazi replied:
Ah, didn't see the conversation about this earlier. Please ignore.

jrajahalme replied:
@christarazi The mock is used from another *_test.go file, and need not be linked into the release code. My understanding is that Go does not automatically exclude dead code.

christarazi commented on pkg/ipcache/ipcache.go:
```diff
@@ -82,7 +82,8 @@ type IPCache struct {
 	listeners []IPIdentityMappingListener
 
-	namedPorts policy.NamedPortMultiMap
+	needNamedPorts bool
```
I think it would be helpful context to briefly explain why `needNamePorts` needs to exist / its role.

aanm replied:
+1 I'm also lost.

jrajahalme replied:
Adding this comment:
```
	// needNamedPorts is initially 'false', but will be changd to 'true' when the
	// clusterwide named port mappings are needed for network policy computation
	// for the first time. This avoids the overhead of maintaining 'namedPorts' map
	// when it is known not to be needed.
	// Both needNamedPorts and namedPorts are protected by 'mutex'.
```

christarazi replied:
Thanks for adding that. Sounds like an optimization. Just out of curiosity, did we measure the impact of this?

christarazi reviewed (approved):
LGTM, a few minor points down below.

jrajahalme committed:
logging: Add rate limiting and use it for named port logging

Rate limit the warning messages emitted to the logs when a port name
can not be (uniquely) mapped to a port number.

Rate limit is set to 1 log every 10 seconds with a burst of 3.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aditighag commented on pkg/endpoint/endpoint.go:
```diff
@@ -225,7 +225,7 @@ type Endpoint struct {
 	// k8sPorts contains container ports associated in the pod.
 	// It is used to enforce k8s network policies with port names.
-	k8sPorts policy.NamedPortsMap
+	k8sPorts policy.NamedPortMap
```
Is there any significance to policy.NamedPortMap vs policy.NamedPortsMap.
You've renamed it in endpoint.go but not in l4.go.

jrajahalme replied:
`policy.NamedPortsMap` is now an interface that abstracts the implementation (simple map, used for ingress that only needs to consider named ports of the POD itself, or a multimap used for egress that needs to map a port name of destination PODs).

christarazi commented on pkg/policy/portmap.go:
```diff
@@ -0,0 +1,170 @@
+	return nil
+}
+
+// NamedPortMap abstracts different maps that implement GetNamedPort method
```
I think it would be helpful to describe why this abstraction is now implemented.

jrajahalme replied:
I'll have to go back and see if this is still needed. I added this in a phase where policy computation was given a map that could be either a regular/simple map (for ingress) or a collated multimap (for egress). Now that the port name resolution is always done via a callback this might not be needed any more.

christarazi commented on pkg/logging/limiter.go:
```diff
@@ -0,0 +1,36 @@
+	}
+}
+
+func (ll Limiter) Allow() bool {
```
Can you provide the Golang docs for these exported types?

christarazi commented on pkg/logging/limiter_test.go:
```diff
@@ -0,0 +1,43 @@
+	c.Assert(limiter.Allow(), Equals, true)
+	c.Assert(limiter.Allow(), Equals, true)
+	c.Assert(limiter.Allow(), Equals, false)
+	c.Assert(limiter.Allow(), Equals, false)
```
It'd be helpful to have a description of what this test is doing and what it expects.

jrajahalme committed:
ipcache: Maintain named ports on demand

Pass Endpoint (or a subset interface) to potential users of Endpoint's
NamedPortsMap so that instead of preparing the map in advance it can
be created on demand in case it is actually needed.

NamedPortsMap is an interface that is implemented by both NamedPortMap
or NamedPortMultiMap, first of which is built from the POD spec, and
the latter is a collection of named ports across all the PODs in the
cluster. When a named port is used in an ingress policy, it can be
translated by using the PODs own NamedPortMap. When a named port is
used in an egress policy the named port can be one of any remote PODs,
so the collection of all the named ports in the cluster is used as the
individual destination POD is not known at policy computation
time. This could be optimized to the scope of PODs sharing the same
set of labels as the destination security IDs are known when policy is
configured.

Currently same port name resolving to different port numbers is
flagged as a runtime error and traffic to this/these ports is not
allowed. To get around this the port names should be made unique so
that each name maps to the same port number and protocol, or
preferably the port names in the egress policies should be replaced
with the desired port numbers.

The maintenance of the collection of named ports across the cluster
described above is started only after first needed for computing the
bpf map entries from an egress policy that uses a port name instead
the number. This way the related overhead is avoided if named ports
are never used in egress policies.

e.k8sPorts member must be accessed with the Endpoint locked. The
normal policy update path already has the endpoint locked, but incremental
selection updates originate from the selector cache without
any Endpoint locks, and therefore we need both locked and unlocked
variant of the GetNamedPortsMap().

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

christarazi commented on pkg/policy/portmap.go:
```diff
@@ -0,0 +1,164 @@
+	ErrUnknownNamedPort     = errors.New("Unknown named port")
+	ErrIncompatibleProtocol = errors.New("Incompatible protocol")
+	ErrNamedPortIsZero      = errors.New("Named port is zero")
+	ErrDuplicateNamedPorts  = errors.New("duplicate named ports")
```
Nit: I think the convention for error strings is to start lower-cased.

jrajahalme replied:
Fixed, thanks!

christarazi commented on pkg/policy/portmap.go:
```diff
@@ -0,0 +1,164 @@
+	ErrDuplicateNamedPorts  = errors.New("duplicate named ports")
+)
+
+type PortProto struct {
```
Could you add go docs for the exported types?

jrajahalme replied:
Added :-)

christarazi reviewed (approved):
LGTM; a few non-blocking comments to address, and should be good to go from my perspective.

jrajahalme committed:
policy: Refactor named ports for centralized logging

Refactor named ports so that error logging can be done in one place.
This makes adding rate limiting in a following commit easier.

Remove the need for policy.NamedPortsMap interface.

Add unit testing.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/logging/limiter.go:
```diff
@@ -0,0 +1,47 @@
+	if ll.bucket == nil {
+		return true // limiter not initialized => no limit
+	}
+	return ll.bucket.Allow()
```
Should we add some metric when this returns false? I think it might be useful to know if we're dropping logs (+ at least the subsystem) and how many / how often.

This might be broader feedback for future usage of this limiter, but it's easier to ensure such capability is present from its first introduction.

jrajahalme replied:
Will have to leave this for a follow-on PR, as other PRs already depend on this one getting merged.

joestringer commented on pkg/policy/portmap.go:
```diff
@@ -0,0 +1,174 @@
+	}
+	pps, ok := npm[name]
+	if !ok {
+		// Return an error the caller can filer out as this happens only for egress policy
```
nit:

```suggestion
		// Return an error the caller can filter out as this happens only for egress policy
```

jrajahalme replied:
Fixed, thanks!

joestringer commented on pkg/policy/l4.go:
```diff
@@ -273,26 +304,31 @@ func ParsePortProtocol(port int, protocol string) (np NamedPort, err error) {
 
-func (npm NamedPortsMap) GetNamedPort(name string, proto uint8) (port uint16, err error) {
+// NamedPortMap abstracts different maps that implement GetNamedPort method
+type NamedPortsMap interface {
```
nit: The comment is not consistent with the name of the interface.

Also, perhaps the `NamedPortMap` implementation should be something like `NamedPortSingleMap` to make it easier to tell the difference & not confuse whether code is referring to the interface or the concrete implementation?

jrajahalme replied:
This interface is now removed in the latest version. Naming changes will have to wait for a folloow-on PR, if still deemed necessary.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1171,6 +1171,14 @@ func (e *Endpoint) SetK8sMetadata(containerPorts []slim_corev1.ContainerPort) er
 
+// GetK8sPorts returns the k8sPorts, which must not be modified by the caller
+func (e *Endpoint) GetK8sPorts() (k8sPorts policy.NamedPortMap) {
+	e.mutex.RLock()
```
I'm not sure I understand the locking here. `pkg/endpoint/lock.go` has lock wrappers that attempt to protect us from various cases by for example propagating back up an error in the case that the endpoint disappears. 

We even have the `unconditionalRLock()` variations, including documentation, which do exactly what this is doing but have conditional comments to describe cases where it's OK to do this vs. `rlockAlive()`. Surely we should be using the wrappers rather than directly interacting with `e.mutex`, even if we decide that the actual logic here is OK because the caller has a reference to the endpoint object?

jrajahalme replied:
Thanks for the reminder, I had completely overlooked this. Fixed.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -963,12 +967,21 @@ func (l4 *L4Policy) AccumulateMapChanges(adds, deletes []identity.NumericIdentit
 		if port == 0 && l4Filter.PortName != "" {
 			var err error
-			port, err = epPolicy.NamedPortsMap.GetNamedPort(l4Filter.PortName, proto)
+			npMap := epPolicy.PolicyOwner.GetNamedPortsMap(direction == trafficdirection.Ingress)
```
nit: performing comparisons to get the bool here to pass into `GetNamedPortsMap()` suggests that the callee should be taking `trafficdirection` directly (or otherwise why do we have a dedicated `trafficdirection` type?).

jrajahalme replied:
Will have to clean this up in a follow-on PR. In this case, without wider refactoring, we'll have to use comparisons somewhere, as trafficdirection is not universally used yet.

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -165,13 +162,15 @@ func (p *EndpointPolicy) computeDirectionL4PolicyMapEntries(l4PolicyMap L4Policy
 			// Fix up the proxy port for entries that need proxy redirection
 			if entry.IsRedirectEntry() {
-				if !lookupDone {
+				if !lookupDone && filter.IsRedirect() {
```
What's the implication of this extra check here? The entry is from `filter.ToMapState()`, so the filter must have a redirect otherwise presumably `entry.isRedirectEntry()` would return false ..?

jrajahalme replied:
This was a result of refactoring a corresponding check out of `LookupRedirectPortLocked()` (called below), but you are right, it is not needed due to filter's entry already being checked on the previous line. Removed :-)

joestringer reviewed (changes requested):
Mostly minor style nits below. One possibly functional piece of feedback at the end.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -40,29 +41,89 @@ import (
+			logfields.PortName:         name,
+			logfields.Protocol:         u8proto.U8proto(proto).String(),
+			logfields.TrafficDirection: "ingress",
+		}).WithError(err).Warning("Skipping named port")
```
Minor followup: This error condition is separate from the exact same message above on line 61 right? It would be helpful to have a way to distinguish the two in case this message is ever logged, so we know which case was hit.

aanm replied:
@jrajahalme will take care of this as a follow up!

joestringer reviewed (approved):
Minor followup comment, otherwise LGTM.
